File: c/TennisGame6.c
#include <stdlib.h>
#include <string.h>
#include <stdio.h>

#include "TennisGame.h"

struct TennisGame {
    int player1Score;
    int player2Score;
    const char *player1Name;
    const char *player2Name;
    char result[18];
};

struct TennisGame *TennisGame_Create(const char *player1Name, const char *player2Name) {
    struct TennisGame *newGame = malloc(sizeof(struct TennisGame));
    newGame->player1Score = 0;
    newGame->player2Score = 0;
    newGame->player1Name = player1Name;
    newGame->player2Name = player2Name;
    return newGame;
}

const char *TennisGame_GetScore(struct TennisGame *game) {
    if (game->player1Score == game->player2Score) {
        // tie score
        char tieScore[18];
        switch (game->player1Score) {
            case 0:
                strcpy(tieScore, "Love-All");
                break;
            case 1:
                strcpy(tieScore, "Fifteen-All");
                break;
            case 2:
                strcpy(tieScore, "Thirty-All");
                break;
            default:
                strcpy(tieScore, "Deuce");
                break;
        }

        strcpy(game->result, tieScore);
    } else if (game->player1Score >= 4 || game->player2Score >= 4) {
        // end-game score
        char endGameScore[18];
        int difference = game->player1Score - game->player2Score;
        if (difference > 2)
            difference = 2;
        switch (difference) {
            case 1:
                sprintf(endGameScore, "Advantage %s", game->player1Name);
                break;
            case -1:
                sprintf(endGameScore, "Advantage %s", game->player2Name);
                break;
            case 2:
                sprintf(endGameScore, "Win for %s", game->player1Name);
                break;
            default:
                sprintf(endGameScore, "Win for %s", game->player2Name);
                break;
        }

        strcpy(game->result, endGameScore);
    } else {
        // regular score
        char regularScore[18];

        char score1[18];
        switch (game->player1Score) {
            case 0:
                strcpy(score1, "Love");
                break;
            case 1:
                strcpy(score1, "Fifteen");
                break;
            case 2:
                strcpy(score1, "Thirty");
                break;
            case 3:
                strcpy(score1, "Forty");
                break;
        };

        char score2[18];
        switch (game->player2Score) {
            case 0:
                strcpy(score2, "Love");
                break;
            case 1:
                strcpy(score2, "Fifteen");
                break;
            case 2:
                strcpy(score2, "Thirty");
                break;
            case 3:
                strcpy(score2, "Forty");
                break;
        };

        sprintf(regularScore, "%s-%s", score1, score2);

        strcpy(game->result, regularScore);
    }

    return game->result;
}

void TennisGame_WonPoint(struct TennisGame *game, const char *playerName) {
    if (strcmp(playerName, "player1") == 0)
        game->player1Score += 1;
    else
        game->player2Score += 1;
}

File: typescript/src/TennisGame6.ts
import { TennisGame } from './TennisGame';

export class TennisGame6 implements TennisGame {
  private player2Name : string;
  private player1Name : string;
  private player1Score: number;
  private player2Score: number;

  constructor(player1Name: string, player2Name: string) {
    this.player1Name = player1Name;
    this.player2Name = player2Name;
    this.player1Score = 0;
    this.player2Score = 0;
  }

  getScore(): string {
    let result: string;

    if (this.player1Score == this.player2Score)
    {
      // tie score
      let tieScore: string;
      switch (this.player1Score)
      {
        case 0:
          tieScore = "Love-All";
          break;
        case 1:
          tieScore = "Fifteen-All";
          break;
        case 2:
          tieScore = "Thirty-All";
          break;
        default:
          tieScore = "Deuce";
          break;
      }

      result = tieScore;
    }
    else if (this.player1Score >= 4 || this.player2Score >= 4)
    {
      // end-game score
      let endGameScore: string;

      if (this.player1Score - this.player2Score == 1) {
        endGameScore = "Advantage " + this.player1Name;
      } else if (this.player1Score - this.player2Score == -1) {
        endGameScore = "Advantage " + this.player2Name;
      } else if (this.player1Score - this.player2Score >= 2) {
        endGameScore = "Win for " + this.player1Name;
      } else {
        endGameScore = "Win for " + this.player2Name;
      }

      result = endGameScore;
    }
    else
    {
      // regular score
      let regularScore: string;

      let score1: string;

      switch (this.player1Score) {
        case 0:
          score1 = "Love";
          break;
        case 1:
          score1 = "Fifteen";
          break;
        case 2:
          score1 = "Thirty";
          break;
        default:
          score1 = "Forty";
          break;
      }

      let score2: string;
      switch (this.player2Score) {
        case 0:
          score2 = "Love";
          break;
        case 1:
          score2 = "Fifteen";
          break;
        case 2:
          score2 = "Thirty";
          break;
        default:
          score2 = "Forty";
          break;
      }

      regularScore = score1 + "-" + score2;

      result = regularScore;
    }

    return result;

  }

  wonPoint(playerName: string): void {
    if (playerName === 'player1')
      this.player1Score += 1;
    else
      this.player2Score += 1;
  }
}
